gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,432 @@
+    if ind == 1:
+        res = 1 - res
+    out_pers, persts = get_persistent_pairs(res, connections, expansion)
+
+    """ Get only the persistent pairs that contribute to
+    the complexity of the decision boundary:
+    These are the ones that are on the different side of the boundary
+    """
+    tmp = out_pers - 0.5
+    cuts = (tmp[:, 0] * tmp[:, 1]) < 0
+
+    goodinds = np.argmin(abs(tmp[cuts, :]), 1)
+    tmp_inds = persts[cuts, :]
+    indices = np.zeros(tmp_inds.shape[0])
+    for i in range(tmp_inds.shape[0]):
+        sele = goodinds[i]
+        indices[i] = tmp_inds[i, sele]
+    # The good values
+    indices = indices.astype(int)
+    coordinates = coords[indices, :].clone().detach().requires_grad_(True)
+    return coordinates
+
+
+class Regularizer:
+    def __init__(self, **kwargs):
+        self.regularizer_arguments = kwargs
+
+    """
+    An abstract class for handling various regularization schemes.
+    A Regularizer needs to have 3 methods:
+    - preprocess
+    - update_params
+    - regularization_penalty
+    """
+
+    @abstractmethod
+    def preprocess(self, **regularizer_arguments: Any) -> Any:
+        """
+        performs preprocessing for the regularizer, if any is necessary
+
+        """
+        return regularizer_arguments
+
+    @abstractmethod
+    def update_params(self, pre_processed_arguments: Any) -> Any:
+        """
+        Any processing needed in a forward pass of the network
+        Anything that doesn't require gradients can go here
+        """
+        return pre_processed_arguments
+
+    @abstractmethod
+    def regularization_penalty(
+        self, forward_passed_arguments: Any, model: torch.nn.Module
+    ) -> Tensor:
+        """
+        Computes the actual regression penalty.
+        This must be an auto-differentiable function of its inputs
+        """
+        return torch.sum(forward_passed_arguments) ** 2
+
+
+class TopologicalRegularizer(Regularizer):
+    """
+    A topological regularizer for a 2-class classifier
+    """
+
+    def preprocess(self, **kwargs):
+        """
+        Initializes the topological regularizer:
+        kwargs:
+            n_neighbors: sizes of the KNN neighbors used for computing the local homology
+            hom_dim: the homologies to be computed
+            expansion:  The size of the clique-expansion when computing homology
+            regularization_dataset: Tensor describing where the linear surrogate function is evaluated
+        """
+
+        defaultKwargs = {
+            "n_neighbors": (6, 20),
+            "hom_dim": (0, 1, 2, 3, 4, 5, 6),
+            "expansion": 6,
+        }
+        self.regularizer_arguments = {**defaultKwargs, **kwargs}
+        print("Preprocessing regularizer")
+        dummy_loader = create_dummy_loader(
+            self.regularizer_arguments["regularization_dataset"]
+        )
+        connections = local_homology_preprocess(
+            self.regularizer_arguments["regularization_dataset"],
```

Comment:
can you please define a `@dataclass` for the regularizer parameters for the topological regularizer where you put all these parameters in? the dataclass does not need to have any method, just store all the needed parameters with type.

Reply:
done, it ended up having `__post_init__` to achieve what was previously known as `preprocess`